fix(trends): Updating Percentage calculation

wmak opened the pull request:
Preventing infinities:
- If the denominator was 0, clickhouse would result in `Inf` or `-Inf`
  which after being cast to None by `handle_data` would mean that top
  trends would end up being several Nulls.
- Using `if` to check the denominator is greater than 0, returning null
  in that case instead of inf/nan

Updating the calculation:
- The result on the frontend will now -1, so that we no longer show 200% when a duration doubles, but show 100%

wmak committed:
fix(trends): Preventing inf from the divides

- If the denominator was 0, clickhouse would result in `Inf` or `-Inf`
  which after being cast to None by `handle_data` would mean that top
  trends would end up being several Nulls.
- Using `if` to check the denominator is greater than 0, returning null
  in that case instead of inf/nan

k-fish commented on src/sentry/api/event_search.py:
```diff
@@ -1549,7 +1549,7 @@ def has_default(self, params):
         "name": "divide",
         "args": [FunctionArg("numerator"), FunctionArg("denominator")],
-        "aggregate": [u"divide", [ArgValue("numerator"), ArgValue("denominator")], None],
+        "transform": "if(greater({denominator},0),divide({numerator},{denominator}),null)",
```
Nice

evanh replied:
Please use `aggregate` over `transform` for these.

evanh reviewed (changes requested):
I don't think we should be defining our own `divide` function that does something different from the clickhouse function. At least name it something else so it's clear it's a custom function.

Also, I'm trying to move away from using raw strings in functions to using the snuba array notation. I know it's not a great format but using raw strings loses out on a lot of the AST optimizations.

I would also add a test that actually does a divide by zero on clickhouse for this case somewhere.

evanh commented on src/sentry/api/event_search.py:
```diff
@@ -1549,7 +1549,7 @@ def has_default(self, params):
     "divide": {
```
Rename

wmak committed:
ref: Rename divide to percentage

- Adding an explicit test of divide by 0
  - Changing user_misery so it doesn't divide by 0
- Moving from transform to aggregate

wmak commented:
Didn't end up moving way from the string syntax since it looks like snuba can't handle the query for ifs [yet](https://gist.github.com/wmak/8d83a02e73a8c033380254b14f617e3a)

Zylphrex commented on src/sentry/api/event_search.py:
```diff
@@ -1546,10 +1546,14 @@ def has_default(self, params):
         "args": [FunctionArg("numerator"), FunctionArg("denominator")],
-        "aggregate": [u"divide", [ArgValue("numerator"), ArgValue("denominator")], None],
+        "aggregate": [
+            u"if(greater({denominator},0),divide({numerator},{denominator}),null)",
```
This is an edge case, and can be done later. If the duration changes from 0ms to 100s between the two time periods, this will be reported as `NULL` and will not be surfaced. This is more problematic with `user_misery` where 0 miserable users in a time period can be common.

wmak committed:
Merge branch 'master' into fix/trends-nan-inf

wmak committed:
fix: Missed a test
